### PR TITLE
UTILS - Logger

### PR DESCRIPTION
**Pull request by kshxtij**
- [x] Added ILogger Interface
- [x] Implement Basic Logger 
- [x] Unit Tests

Verbosity levels are defined based on the ros and roscpp spec found [here](http://wiki.ros.org/Verbosity%20Levels).

While in a Ros node enviroment, we'd pass in RosLogger into the classes which implements logging to stdout, stderr, text file as well as /rosout (a ros topic that runs centrally and collects all logging message). RosCpp defines these using macros within the standard ros library so RosLogger would just wrap the ros provided macro/definitions.

### CODE REVIEW

File: lib/io/gpio.cpp
Comment (mifrandir) [inline lib/io/gpio.cpp:5]: I dont like that this is a pointer. The logger wont hold a lot of data, right? So passing it by value should be fine.

File: lib/utils/logger.cpp
Comment (mifrandir) [inline lib/utils/logger.cpp:21]: Way too many levels to be practical. At most have fatal, warnings, info, and debug.

**Review by mifrandir — CHANGES_REQUESTED** [review]
Maybe logging should live in core? This seems pretty essential.

It also seems like a step back to wrap macros in function calls but I dont suppose there's any way around this without linking against ROS?

File: lib/utils/logger.cpp
Comment (kshxtij) [reply lib/utils/logger.cpp:21]: Yeah that's what i thought too, i just followed ROS's levels but i'll remove kError.

kNone is there for tests when we don't want output to stdout/stderr

**Comment by kshxtij** [comment]
I think ROS's logging utility is too useful to not link against especially when it has the ability to centralise logging as well, I agree that wrapping macros with functions isn't ideal so i'll look into calling ros's logger directly.

**Comment by kshxtij** [comment]
The only logging api exposed by ROS is through the macros, while it isn't ideal to wrap them with functions i think that's the most realistic way to do it.  Especially since we definitely want to be using RosLogger within a multi-node, multi-controller setup.

It also directly links with some other ros extensions we might want to use in the future.

**Comment by kshxtij** [comment]
As soon as this PR is reviewed and merged, i'll go through the preexisting code as well as currently open PRs to add logging where needed (like the comments in adc/i2c by Ishan)

File: lib/core/logger.cpp
Comment (TomLonergan03) [inline lib/core/logger.cpp:98]: What's the point of kError as a level? It only seems to be used here and isn't accessible through `intToLevel`.
Could it be removed and just use kFatal instead?

File: lib/core/logger.cpp
Comment (TomLonergan03) [inline lib/core/logger.cpp:64]: Is there a reason for doing `va_list args`, `va_start` etc. in each logging level instead of within print itself?

**Review by TomLonergan03 — CHANGES_REQUESTED** [review]

File: lib/core/logger.cpp
Comment (kshxtij) [reply lib/core/logger.cpp:98]: Forgot to remove it from the enum and these functions when i deleted log.error()

File: lib/core/logger.cpp
Comment (kshxtij) [reply lib/core/logger.cpp:64]: Can only use `va_start` within a function that accepts `...` (or ellipsis) as a parameter

Comment (TomLonergan03) [reply lib/core/logger.cpp:64]: makes sense

**Review by TomLonergan03 — APPROVED** [review]

File: lib/core/logger.cpp
Comment (maxguy2001) [inline lib/core/logger.cpp:35]: Can parts of this function not be `const`? Also naming in this function?

File: lib/core/logger.cpp
Comment (mifrandir) [inline lib/core/logger.cpp:38]: This is not how we usually get timestamps, right? This should use a time source instead. This might make it deterministic in testing, too?

File: lib/core/logger.cpp
Comment (mifrandir) [inline lib/core/logger.cpp:42]: What is  this `print_micro` good for?

File: lib/core/logger.cpp
Comment (mifrandir) [inline lib/core/logger.cpp:39]: In general, I'm not a fan of the `printf` style. I know there are some merits in terms of performance but I'm not sure that's worth it?

File: lib/core/logger.cpp
Comment (mifrandir) [inline lib/core/logger.cpp:95]: All of these functions look identical. Why not have a private `void log(const std::string label, const Level level, std::ofstream output_stream)`?

The strings can be static in their parent calls so don't worry about allocations.

File: lib/core/logger.hpp
Comment (mifrandir) [inline lib/core/logger.hpp:13]: Do we really want constructors here?

**Review by mifrandir — CHANGES_REQUESTED** [review]

File: lib/core/logger.hpp
Comment (kshxtij) [reply lib/core/logger.hpp:13]: I think i did this because module_ and level_ are members of ILogger and not Logger so to set them i'd need to defer to ILoggers constructor?

File: lib/core/logger.cpp
Comment (kshxtij) [reply lib/core/logger.cpp:95]: You'd need to have the 

```cpp
    va_list args;
    va_start(args, format);
    
    va_end args;
```

within the `fatal`, `info`, `debug`, `warn` functions due to the ellipsis parameters to accept multiple arguments so i don't see what you could move to another function

File: lib/core/logger.cpp
Comment (kshxtij) [reply lib/core/logger.cpp:39]: I don't think this is a problem because we use printf style args within the logging functions themselves unless we replace everything with std::format and Variadic templates but im not sure that's needed

File: lib/core/logger.cpp
Comment (kshxtij) [reply lib/core/logger.cpp:38]: Will switch over to the constructor accepting a ITimeSource and using now().

File: lib/core/logger.hpp
Comment (kshxtij) [reply lib/core/logger.hpp:13]: Changed the location of module_ and level_ to Logger instead of ILogger

File: lib/core/logger.cpp
Comment (SnickeyX) [inline lib/core/logger.cpp:42]: Is this supposed to always be true?

**Review by SnickeyX — CHANGES_REQUESTED** [review]

File: lib/core/logger.cpp
Comment (kshxtij) [reply lib/core/logger.cpp]: Removed, was an artifact of last year

**Review by SnickeyX — CHANGES_REQUESTED** [review]

File: test/core/logger.cpp
Comment (mifrandir) [inline test/core/logger.cpp:8]: `const`


`use_stdout`

Comment (mifrandir) [reply test/core/logger.cpp:8]: Also, maybe call it `testLog`?

File: test/core/logger.cpp
Comment (mifrandir) [inline test/core/logger.cpp:33]: Maybe combine these into one test? Just have subsequent calls to logHelper.

File: lib/core/logger.hpp
Comment (mifrandir) [inline lib/core/logger.hpp:20]: module

File: lib/core/logger.hpp
Comment (mifrandir) [inline lib/core/logger.hpp:27]: const

File: lib/core/logger.hpp
Comment (mifrandir) [inline lib/core/logger.hpp:28]: `time_source`

File: lib/core/logger.hpp
Comment (mifrandir) [inline lib/core/logger.hpp:15]: `=0`

also check the ROS interface but it may now be possible to put a `const` at the end here, too.

**Review by mifrandir — CHANGES_REQUESTED** [review]

**Review by TomLonergan03 — APPROVED** [review]
This is much nicer having ditched all the duplicate functions

File: lib/core/logger.cpp
Comment (SnickeyX) [inline lib/core/logger.cpp:19]: naming

**Review by SnickeyX — CHANGES_REQUESTED** [review]

**Review by SnickeyX — APPROVED** [review]

File: lib/core/logger.cpp
Comment (maxguy2001) [inline lib/core/logger.cpp:54]: ...

File: lib/core/logger.cpp
Comment (maxguy2001) [inline lib/core/logger.cpp:22]: :(

File: lib/core/logger.cpp
Comment (mifrandir) [inline lib/core/logger.cpp:17]: Identifiers with suffix `_t` usually refer to a type.

**Review by maxguy2001 — APPROVED** [review]
Builds

**Review by mifrandir — APPROVED** [review]

**Pull request merged**
